tensorflow/compiler/xla/service/algebraic_simplifier.cc

```diff
@@ -1665,6 +1668,228 @@ StatusOr<HloInstruction*> AlgebraicSimplifierVisitor::OptimizeDotOfGather(
+
+  // Transform rhs.
+  // Let's walk through an example to see how rhs transformation is done and the
+  // meaning of some variables used in the implementation. Say we have the
+  // following original Hlo IR:
+  //
+  //   t0 = F32[2, 2, 3] parameter(0)
+  //   t1 = F32[2, 3, 2] transpose(t0) dimensions={0, 2, 1}
+  //   t2 = F32[2, 6] reshape(t1)
+  //   t3 = F32[6, 2] constant(...)
+  //   dot = F32[2, 2] dot(t2, t3) lhs_contracting_dims={1},
+  //                               rhs_contracting_dims={0}
+  //
+  // 1) The sorted lhs_contracting_dims is {1, 2}. We compute t1's subshape
+  //    within the lhs_contracting_dims: reshape_subshape = [3, 2].
+  // 2) rhs_contracting_dims is {0}, t3's shape is [6, 2]. We replace dim 0 of
+  //    t3's shape by reshape_subshape. Now we get inverted reshape's shape:
+  //    inverted_reshape_shape = [3, 2, 2]. We create inverted reshape op:
+  //      inverted_reshape = F32[3, 2, 2] reshape (F32[6, 2] t3)
+  //    and update rhs_contracting_dims to {0, 1}.
+  // 3) We compute the inverse lhs transpose dims
+  //    inverse_transpose_dims = {0, 2, 1}. The lhs contracting dims part is
+  //    inverse_transpose_sub_dims = {2, 1}.
+  // 4) For each element in inverse_transpose_sub_dims, we subtract the start
+  //    position of lhs_contracting_dims, 1 in this example, then add the start
+  //    position of rhs_contracting_dims, 0 in this example. Now we get the
+  //    actual transpose dims within rhs_contracting_dims:
+  //    rhs_transpose_sub_dims = {1, 0}.
+  // 5) We compute the transpose dims of the inverted transpose:
+  //    inverted_transpose_dims = {1, 0, 2}, and then create inverted transpose:
+  //      inverted_transpose =
+  //         F32[2, 3, 2] transpose(inverted_reshape) dimensions = {1, 0, 2}
+
+  // Invert reshape.
+  std::vector<int64> reshape_subshape;
+  reshape_subshape.reserve(lhs_contracting_dims.size());
+  for (auto dim : lhs_contracting_dims) {
+    reshape_subshape.push_back(reshape->operand(0)->shape().dimensions(dim));
+  }
+  CHECK(rhs_contracting_dims.size() == 1);
+  auto inverted_reshape_shape = rhs->shape().dimensions();
+  inverted_reshape_shape.erase(inverted_reshape_shape.begin() +
+                               rhs_contracting_dims[0]);
+  inverted_reshape_shape.insert(
+      inverted_reshape_shape.begin() + rhs_contracting_dims[0],
+      reshape_subshape.begin(), reshape_subshape.end());
+  HloInstruction* inverted_reshape =
+      computation_->AddInstruction(HloInstruction::CreateReshape(
+          ShapeUtil::MakeShape(constant->shape().element_type(),
+                               inverted_reshape_shape),
+          constant));
+  // The inverted reshape "unsquishes" the single rhs contracting dim into
+  // multiple dims.
+  for (int64 i = 1; i < lhs_contracting_dims.size(); ++i) {
+    rhs_contracting_dims.Add(rhs_contracting_dims[0] + i);
+  }
+  rhs = inverted_reshape;
+
+  // Invert transpose.
```

Comment:
It seems like doing the inverse transpose should be a similarly straightforward operation?

Reply:
Getting the shape for the inverse transpose should be similar as your previous comment about reshape.  But to create the inverse transpose op, we also need to compute the transpose dimensions of the inverse transpose. I can not think of a way to avoid subtracting lhs_contracting_dims[0] and adding back rhs_contracting_dims[0] for each rhs contracting dims as in the current implementation.

Reply:
Do you therefore think that the relative simplicity of these two approach is the same?